File: GbxTools3D/GbxTools3D.Client/Modules/Animation.cs
﻿using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;

namespace GbxTools3D.Client.Modules;

[SupportedOSPlatform("browser")]
internal static partial class Animation
{
    [JSImport("createPositionTrack", nameof(Animation))]
    public static partial JSObject CreatePositionTrack(double[] times, double[] values, bool discrete = false);

    [JSImport("createQuaternionTrack", nameof(Animation))]
    public static partial JSObject CreateQuaternionTrack(double[] times, double[] values, bool discrete = false);

    [JSImport("createRotationXTrack", nameof(Animation))]
    public static partial JSObject CreateRotationXTrack(double[] times, double[] values);

    [JSImport("createRotationYTrack", nameof(Animation))]
    public static partial JSObject CreateRotationYTrack(double[] times, double[] values, bool discrete = false);

    [JSImport("createRelativePositionYTrack", nameof(Animation))]
    public static partial JSObject CreateRelativePositionYTrack(double[] times, double[] values, JSObject referenceObj, bool discrete = false);

    [JSImport("createClip", nameof(Animation))]
    public static partial JSObject CreateClip(string name, double duration, JSObject[] tracks);

    [JSImport("createMixer", nameof(Animation))]
    public static partial JSObject CreateMixer(JSObject object3D);

    [JSImport("createAction", nameof(Animation))]
    public static partial JSObject CreateAction(JSObject mixer, JSObject clip, JSObject? object3D = null);

    [JSImport("playAction", nameof(Animation))]
    public static partial JSObject PlayAction(JSObject action);

    [JSImport("pauseAction", nameof(Animation))]
    public static partial JSObject PauseAction(JSObject action);

    [JSImport("resumeAction", nameof(Animation))]
    public static partial JSObject ResumeAction(JSObject action);

    [JSImport("playMixer", nameof(Animation))]
    public static partial void PlayMixer();

    [JSImport("pauseMixer", nameof(Animation))]
    public static partial void PauseMixer();

    [JSImport("setMixerTime", nameof(Animation))]
    public static partial void SetMixerTime(double time);

    [JSImport("setMixerTimeScale", nameof(Animation))]
    public static partial void SetMixerTimeScale(double timeScale, bool isPaused);

    [JSImport("disposeMixers", nameof(Animation))]
    public static partial void DisposeMixers();
}

File: GbxTools3D/GbxTools3D.Client/wwwroot/js/animation.js
import * as THREE from 'three';

let dotNetObjRef;
let mixers = [];
let mixerTimeScale = 1;

export function createPositionTrack(times, values, discrete) {
    return new THREE.VectorKeyframeTrack('.position', times, values, discrete ? THREE.InterpolateDiscrete : THREE.InterpolateLinear);
}

export function createQuaternionTrack(times, values, discrete) {
    return new THREE.QuaternionKeyframeTrack('.quaternion', times, values, discrete ? THREE.InterpolateDiscrete : THREE.InterpolateLinear);
}

export function createRotationXTrack(times, values) {
    return new THREE.NumberKeyframeTrack('.rotation[x]', times, values);
}

export function createRotationYTrack(times, values) {
    return new THREE.NumberKeyframeTrack('.rotation[y]', times, values);
}

export function createRelativePositionYTrack(times, values, referenceObj) {
    let refY = referenceObj.position.y;
    for (let i = 0; i < values.length; i++) {
        values[i] = refY - values[i];
    }
    return new THREE.NumberKeyframeTrack('.position[y]', times, values);
}

export function createClip(name, duration, tracks) {
    return new THREE.AnimationClip(name, duration, tracks);
}

export function createMixer(object) {
    const mixer = new THREE.AnimationMixer(object);
    mixer.timeScale = 0;
    mixers.push(mixer);
    return mixer;
}

export function createAction(mixer, clip, obj) {
    return mixer.clipAction(clip, obj);
}

export function playAction(action) {
    action.play();
}

export function pauseAction(action) {
    action.paused = true;
}

export function resumeAction(action) {
    action.paused = false;
}

export function playMixer() {
    for (let i = 0; i < mixers.length; i++) {
        mixers[i].timeScale = mixerTimeScale;
    }
}

export function pauseMixer() {
    for (let i = 0; i < mixers.length; i++) {
        mixers[i].timeScale = 0;
    }
}

export function setMixerTime(time) {
    for (let i = 0; i < mixers.length; i++) {
        const mixer = mixers[i];
        let prevTimeScale = mixer.timeScale;
        mixer.timeScale = 1;
        mixer.setTime(time);
        mixer.timeScale = prevTimeScale;
    }
}

export function setMixerTimeScale(timeScale, isPaused) {
    mixerTimeScale = timeScale;

    if (!isPaused) {
        for (let i = 0; i < mixers.length; i++) {
            const mixer = mixers[i];
            mixer.timeScale = timeScale;
        }
    }
}

let prevMixerTime;

export function updateMixer(delta) {
    for (let i = 0; i < mixers.length; i++) {
        const mixer = mixers[i];
        mixer.update(delta);
        if (mixer.time !== prevMixerTime) {
            dotNetObjRef.invokeMethodAsync("UpdateTimeline", mixer.time);
            prevMixerTime = mixer.time;
        }
    }
}

export function disposeMixers() {
    for (let i = 0; i < mixers.length; i++) {
        mixers[i].stopAllAction();
    }
    mixers = [];
    prevMixerTime = null;
}

export function registerDotNet(reference) {
    dotNetObjRef = reference;
}